Added tutorial page on log trace in application (#80)
MicroEJ/docs: Tutorials/tutorialInstrumentJavaCodeForLogging.rst

      public static void switchState(ApplicationState newState) {
         tracer.recordEvent(EVENT_ID);

         previousState = currentState;
         currentState = newState;

         tracer.recordEventEnd(EVENT_ID);
      }
   
   The `Tracer <https://repository.microej.com/javadoc/microej_5.x/apis/ej/trace/Tracer.html>`_ object records the entry/exit of method ``switchState`` with event ID ``0``.
   
This produces the following output:

.. code-block::

   [TRACE: Application] Event 0x0()
   [TRACE: Application] Event End 0x0()

.. note::

   The default Platform implementation of the ``Trace`` library prints the events to the console.
   See :ref:`trace_implementations` for other available implementations such as `SEGGER SystemView <https://www.segger.com/products/development-tools/systemview/>`_ tool.
   
Log with the Message Library
----------------------------

The library `ej.library.runtime.message <https://repository.microej.com/artifacts/ej/library/runtime/message/>`_ was designed to enable logging while minimizing RAM/ROM footprint and CPU usage.
For that, logs are based on message identifiers, which are stored on integers instead of using of constant Strings.

> An external documentation must be maintained to describe each message identifier and its expected arguments.
> Combined with the category, the printed message identifier value allows the user to find the corresponding error/warning/info description.

Reviewer: `the category` is introduced here without context and it can't be understood without reading the rest of the section.
Here is a suggested formulation:
```suggestion
In addition to a message identifier, the category of the message allows the user to find the corresponding error/warning/info description.
An external documentation must be maintained to describe all message identifiers and their expected arguments for each category.
```

       The `LogRecord <https://repository.microej.com/javadoc/microej_5.x/apis/java/util/logging/LogRecord.html>`_ object wraps the String message and the log level. 
    -  Log level-specific methods, like ``severe(String msg)``, that call the aforementioned ``log(...)`` method with the correct level.

-  The library defines a default `Handler <https://repository.microej.com/javadoc/microej_5.x/apis/java/util/logging/Handler.html>`_ implementation, called `DefaultHandler <https://repository.microej.com/javadoc/microej_5.x/apis/ej/util/logging/handler/DefaultHandler.html>`_,
   that prints the message of the `LogRecord <https://repository.microej.com/javadoc/microej_5.x/apis/java/util/logging/LogRecord.html>`_ on the standard error output stream.
   It also prints the stack trace of the `Throwable <https://repository.microej.com/javadoc/microej_5.x/apis/java/lang/Throwable.html>`_ associated with the `LogRecord <https://repository.microej.com/javadoc/microej_5.x/apis/java/util/logging/LogRecord.html>`_ if there is one.

Here is a short example of how to use this library to log the entry/exit of the ``switchState()`` method:

> #. Add the following dependency to the ``module.ivy``: 

Reviewer: Because of the nested list (the logger API provides....: - item 1 - item 2), the fonts of this list items are larger than the lists in the other sections.  I find it a bit confusing. 
I couldn't find a workaround except by reformulating the item to remove the sub-list.

Reviewer 2: fixed

   .. code-block:: xml

      <dependency org="ej.library.eclasspath" name="logging" rev="1.1.0"/>

#. Call the logging API to log some info text:

   .. code-block:: java
      :emphasize-lines: 5,6,7
     
      public static void switchState(ApplicationState newState) {
         previousState = currentState;
         currentState = newState;

         Logger logger = Logger.getLogger(Main.class.getName());
         logger.log(Level.INFO, "The application state has changed from " + previousState.toString() + " to "
               + currentState.toString() + ".");
      }

This produces the following output: 

.. code-block::
      
   main INFO: The application state has changed from UNINSTALLED to INSTALLED.

.. note::

>    Unlike the two other libraries discussed here, the ``Logging`` library is entirely based on Strings (log IDs and messages). 
>    Note that String operations can lead to performance issues and that String objects use significant ROM space. 
>    When possible, prefer using a logging solution that uses primitive types over Strings.

Reviewer: ```suggestion
   Unlike the two other libraries discussed here, the ``Logging`` library is entirely based on Strings (log IDs and messages). 
   String operations can lead to performance issues and String objects use significant ROM space. 
   When possible, prefer using a logging solution that uses primitive types over Strings.
```

The ``Note that`` feels redundant as we are already inside a `.. note::` item.

Remove Logging Related Code
---------------------------

This section describes some techniques to remove logging related code, which saves memory footprint when logging is disabled at runtime.
This is typically useful when building two Firmware flavors: one for production and one for debug.

Wrap with a Constant If Statement
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A boolean :ref:`constant <section.classpath.elements.constants>` declared in an ``if`` statement can be used to fully remove portions of code. 
When this boolean constant is detected to be ``false``, the wrapped code becomes unreachable and is not embedded.

.. note::
    More information about the usage of constants and ``if`` code removal can be found in the :ref:`Classpath <if_constant_removal>` section.

#. Let's consider a constant ``com.mycompany.logging`` declared as ``false`` in a resource file named ``example.constants.list``.

    .. image:: images/tuto_microej_trace_constant.png

>         :align: center

Reviewer: This tip is useful but what's the best practice regarding how to do it in practice to produce both a production and a debug firmware?

- 2 projects? 1 for production and 1 for debug with various properties set in each project
- can it be set in the launcher? how?

Reviewer 2: 2 projects or 1 project with a custom module.ant that generates prod/debug options properties files depending on a build option - OK I understand you point I will create a dedicated ticket how to build multiple firmware flavors from build options.
